### PR TITLE
Milestone: Friendships V2

### PR DESCRIPTION
**Pull request by Janluc**
## Social Media Milestone
- Add the ability to view mutual friends of other users registered to the website.

### CODE REVIEW

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Changes Required! :recycle:
### Hi @Janluc and @deikdesign :clap: :clap:
#### You have done a good job so far :+1: However, some requirements are missing/not implemented correctly. Check my comments below :point_down:

#### 1) App Behaviour:
 - when friendship is confirmed, an additional row with reversed user-friend pair should be created
Screenshot from 2021-03-05 19-21-03
 - The app should display posts in a timeline (user posts + posts from their friends). Currently, it's showing all the posts regardless of whether the user and the post author are friends.

#### 2) Test:
 - Add integration tests.

#### 3) Readme:
 - Improve the setup section instruction: add instruction to clone the repo
 - Add instruction on how to run the test.

#### 4) View & Logic:
 - Remove all the logic from the view.

#### Optional:
 - Solve the N+1 problem to improve the queries that fetch data from the database.

#### _Reviewed by Temitayo Adeleke_

**Comment by Janluc** [comment]
## Reversed user-friend pair already implemented. Please refer to the Users model in the repo.

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -2,6 +2,14 @@
...
+      <h3>Friends in common:</h3>
+      <ul>
+        <% @mutual_friends.each do |friend|%>
+          <li><%=friend.name%></li>
+        <% end %>
+      </ul>
+    <% end %>
     <% if current_user.friend?(@user) %>
       <p>Already Friends!</p>
     <% else %>
```
Comment (voscarmv) [inline app/views/users/show.html.erb]: Please move this to a helper method.

File: app/views/posts/_post.html.erb [hunk 2]
```diff
@@ -1,4 +1,5 @@
-<li class="post">
+<% if current_user.friends.include?(post.user) or post.user == current_user %>
```
Comment (voscarmv) [inline app/views/posts/_post.html.erb]: Please move this to a helper method and call it from there.

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -2,6 +2,14 @@
...
   <br>
   <% unless current_user == @user %>
+    <% if @mutual_friends.any? %>
+      <h3>Friends in common:</h3>
+      <ul>
+        <% @mutual_friends.each do |friend|%>
+          <li><%=friend.name%></li>
+        <% end %>
+      </ul>
+    <% end %>
```
Comment (voscarmv) [inline app/views/users/show.html.erb]: Please move this to a helper method.

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Impressive work. 💻 👀 ✨ 
- Great job applying some of the changes requested previously.
## However, there are some changes required 👀 🧰 🔨 🔧 
- The point of adding two rows per friendship is to avoid having to construct the list of friends with a double query as you do here:
image
- Please refer to these articles to get a better idea of what's expected here:
https://explainextended.com/2009/03/07/selecting-friends/
https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table
- Basically what you have to do is take the queries from the articles and translate them to Active Record queries. Remember: you must avoid using more than a single query to pull a user's complete list of friends. Otherwise there is no real efficiency improvement in using double friendshp rows.
- As the previous Code Reviewer mentioned: please move all logic from the views into helper methods.

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Comment by Janluc** [comment]
### Thank you for the explanation on the inverse table I was a bit lost on how it was to be implemented.

**Review by mke2111 — CHANGES_REQUESTED** [review]
# **STATUS: CHANGES REQUIRED 🔴 ⚠️** 

## Hey @Janluc & @deikdesign, 👋🏽 
 **Thank you for the good job you got going on and for making the requested changes.** We have one more thing to do.

## REQUIRED CHANGES 🔴 

- [ ] Please move all the logic from the views folder to the helpers folder. This includes all conditional statements and loops. 

**Kindly make all the requested changes and submit another CRR. 👍🏽** 

## Cheers!! 🍹 & Happy Coding.

# Reach me on slack for a chitchat 🆘  **Roy Mukuye**

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @deikdesign and @Janluc 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed.

##

Kindly work on the suggested changes 👇.
- [ ] MIT license is missing from the readme file.
- [ ] Please add `.vscode` to `.gitignore`.
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Review by umairarshadbutt — APPROVED** [review]
# **Status: PROJECT APPROVED** :champagne:
Excellent job on this project guys. :fireworks: Keep it up.
##
**All the required tasks have been fulfilled in this project. I will go ahead and mark this as complete.**
- Congratulations on the splendid job @deikdesign and @Janluc ! :+1: 
- You can now merge to master and submit the project completion form :rocket: . I wish you all the very best in your next tasks :1st_place_medal:
##
**If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Pull request merged**